### PR TITLE
Adds Product Manager

### LINKED ISSUE #7: UI for creating, changing, and deactivating a product
Only accessible to Board

Needs new endpoint on the backend

### PR DESCRIPTION
**Pull request by A-Guldborg**
Closes #7 
Enables adding and updating products through the Shifty app.

Note, a couple things to be discussed:

- With the new endpoint in https://github.com/AnalogIO/analog-core/pull/237 we can now see non-visible products and the property as well. We should figure out some styling / way to clearly indicate which products are not visible (i.e. likely disabled permanently). Right now, it is a gray-ish row on non-visible products, as well as a switch that indicates it is not visible when in edit mode.
- Having a dropdown of checkboxes for the user groups on a product contains an error: We are not able to override the custom text function in the dropdown in the edit form - likely because the editform itself is auto-generated. Now the solution instead is a checkbox for each usergroup but not in a dropdown. It is possible that we can add this to a dropdown list (but not a "select" object) and thus custom-implement the dropdown we want. But that is not very blazor-ish.

Please give inputs to the above discussion points and/or anything else :)

~~NB: Blocked until https://github.com/AnalogIO/analog-core/pull/237 is merged as we require the endpoint created in that PR.~~

image
image

### CODE REVIEW

**Comment by A-Guldborg** [comment]
Almost done.

Few things:
1. I can't for the love of God format the text on the multi-select dropdown. Right now, when editing a produc,t, it shows only one value, e.g. "Customer", despite have both Customer and Board as user groups.
2. It seems the PR https://github.com/AnalogIO/analog-core/pull/223 does not work as expected. It does not expose the product user groups in the API on a get request, but I believe that might just be one single line to be added in `coffeecard\CoffeeCard.WebApi\Controllers\v2\ProductsController.cs` in method `MapProductToDto`:
```
AllowedUserGroups = product.ProductUserGroup.Select(e => e.UserGroup)
```

(But I will need to test that before creating such a PR). Once at least the second issue mentioned here is fixed, this PR is ready for review.

File: Shifty.App/Components/ProductManager.razor [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+            EndIcon="@Icons.Material.Outlined.Add" 
+            OnClick="@AddItem">
+                Add Product
+            </MudButton>
+    </MudToolBar>
+</MudPaper>
+@code
+{
+    [Parameter]
+    public System.Security.Claims.ClaimsPrincipal User { get; set; }
```
[the remark this answers is not in the document]
Comment (A-Guldborg) [reply Shifty.App/Components/ProductManager.razor:91]: Nothing, so I shall have it removed

File: Shifty.App/Services/IProductService.cs [hunk 2]
```diff
@@ -12,6 +13,8 @@ public interface IProductService
         /// Gives products available to user
         /// </summary>
         /// <returns>Collection of products in the form of ProductDtos. The Collection is null if an error happens.</returns>
-        Task<Try<IEnumerable<ProductDto>>> GetProducts();
+        Task<Try<IEnumerable<ProductResponse>>> GetProducts();
+        Task<Try<ChangedProductResponse>> UpdateProduct(UpdateProductRequest product);
```
[the remark this answers is not in the document]
Comment (A-Guldborg) [reply Shifty.App/Services/IProductService.cs:23]: If you say so ┻━┻ ︵ ＼( °□° )／ ︵ ┻━┻

**Review by marfavi — CHANGES_REQUESTED** [review]
I request all changes in #22 to be merged before continuing, which fixes the horizontal scrolling in the table.

And I renamed your PR :)

**Review by marfavi — APPROVED** [review]
OK

**Pull request merged**